### PR TITLE
support user policy submission to bone

### PR DESCRIPTION
**Pull request by starsy**

### CODE REVIEW

File: alarm/AlarmManager2.js [hunk 1]
```diff
@@ -520,19 +520,23 @@ module.exports = class {
           type = "mac";
           target = alarm["p.device.mac"];
           break;
-        default:
+          case "ALARM_VIDEO":
+            type = "domain";
+            target = alarm["p.dest.id"];
+            break;
+          default:
```
[the remark this answers is not in the document]
Comment (MelvinTo) [reply alarm/AlarmManager2.js]: i don't understand why specifically adding code for ALARM_VIDEO here.

File: alarm/AlarmManager2.js [hunk 2]
```diff
@@ -520,19 +520,23 @@ module.exports = class {
...
           type = "ip";
           target = alarm["p.dest.ip"];
 
           if(alarmInfo && alarmInfo.type === "dns") {
-            type = "dns"
+            type = "dns";
```
Comment (MelvinTo) [inline alarm/AlarmManager2.js]: i don't like comma ;)

File: alarm/ExceptionManager.js [hunk 3]
```diff
@@ -164,6 +177,9 @@ module.exports = class {
           callback(err);
         });
       });
+
+      Bone.submitException('allow', exception);
```
Comment (MelvinTo) [inline alarm/ExceptionManager.js]: Shall we add exception to cloud only after exception is added to local redis successfully?

File: package.json [hunk 4]
```diff
@@ -19,20 +19,33 @@
   "dependencies": {
```
[the remark this answers is not in the document]
Comment (MelvinTo) [reply package.json]: i think it's just a reorder, anyway, only the "script" section of this package.json is used (by unit-testing)

File: alarm/AlarmManager2.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (starsy) [reply alarm/AlarmManager2.js]: this is to address the issue caused by a wrong data in sample alarms for unit test, will fix the data there and remove these code

File: config/features.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+let log = require("../net2/logger.js")(__filename, "info");
+
+let fs = require('fs');
+let f = require('../net2/Firewalla.js');
+
+let features = JSON.parse(fs.readFileSync(f.getFirewallaHome() + "/config/features.json", 'utf8'));
```
[the remark this answers is not in the document]
Comment (starsy) [reply config/features.js]: thinking adding a general feature switches for future, config.json seems not that fit for the requirement. this file can be controlled by upgrade. to do it at checkin is a good idea, may enhance it later.

File: alarm/AlarmManager2.js (hunk 2 above)
Comment (starsy) [reply alarm/AlarmManager2.js]: webstorm likes it.

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (starsy) [reply package.json]: webstorm does auto reorder. also didn't know the package was not managed here as a standard way. exploring the code and guessing what is for what is like blindly walking in a dark maze w/ lots of pitfalls...

File: alarm/ExceptionManager.js (hunk 3 above)
Comment (starsy) [reply alarm/ExceptionManager.js]: no need, this only captures user's intention, no matter it is locally stored successfully or not.

File: lib/Bone.js [hunk 6]
```diff
@@ -629,3 +631,83 @@ exports.getServiceConfig = function(callback) {
     }
   });
 };
+
+exports.submitException = function(action, exception) {
+  log.info("Submit User Exception: '%s' => ", action, exception, {});
```
[the remark this answers is not in the document]
Comment (starsy) [reply lib/Bone.js]: the collection will be anonymous in the cloud to keep privacy, so no hash should be ok here.

File: lib/Bone.js [hunk 7]
```diff
@@ -629,3 +631,83 @@ exports.getServiceConfig = function(callback) {
...
+  callback = callback || function() {};
+
+  if (!features.isOn("user:policy:submit")) {
+    log.info("User policy submit feature is off", {});
+    callback(null, false);
+    return;
+  }
+
+  log.info("Pushing user policy to cloud: ", policy);
+  let url = getEndpoint() + '/userpolicy';
```
[the remark this answers is not in the document]
Comment (starsy) [reply lib/Bone.js]: no existing endpoint match against this purpose, or you can name it.

File: lib/Bone.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (jerrchen) [reply lib/Bone.js]: that argument doesn't work.  use hash please.  We take privacy very seriously.

File: lib/Bone.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (jerrchen) [reply lib/Bone.js]: use something like this.

bone.intel(alarm['p.dest.ip'],"autoblock",alarm,(err)=>{});

File: config/features.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jerrchen) [reply config/features.js]: Lets add it to config.json, it already have melvin's *sensors*, which is already a set of features.   You can call something else, but let's keep it unified.

File: alarm/PolicyManager2.js [hunk 8]
```diff
@@ -219,6 +224,12 @@ class PolicyManager2 {
           return Promise.reject("policy " + policyID + " doesn't exists");
         }
 
+        this.getPolicy(policyID).then((policy) => {
+          Bone.submitPolicy('unblock', policy);
+        }).catch((err) => {
+          // null
+        });
+
```
Comment (jerrchen) [inline alarm/PolicyManager2.js]: use bone.intel API

**Comment by starsy** [comment]
PR Updated @ Oct 30

**Pull request merged**